File: src/conversation-parser.ts
import { readFileSync, statSync } from 'node:fs';
import type { ConversationExchange, ToolCallInfo } from './types.js';

// --- Constants ---

/** Skip files larger than 20MB */
export const MAX_FILE_BYTES = 20 * 1024 * 1024;

/** Record types to skip entirely (no useful text) */
const SKIP_TYPES = new Set([
  'progress',
  'queue-operation',
  'file-history-snapshot',
]);

/** Content block types to skip within assistant messages */
const SKIP_BLOCK_TYPES = new Set([
  'tool_use',
  'tool_result',
  'thinking',
]);

// --- Types ---

interface JsonlRecord {
  type?: string;
  sessionId?: string;
  cwd?: string;
  timestamp?: string;
  message?: {
    role?: string;
    content?: string | ContentBlock[];
  };
}

interface ContentBlock {
  type: string;
  text?: string;
}

// --- Parser ---

/**
 * Parse a conversation archive JSONL file and extract user/assistant text
 * into structured markdown suitable for chunking and embedding.
 *
 * @param absolutePath - Full path to the .jsonl file
 * @param projectDir  - The project directory name (e.g. "-Users-jdoe-my-project")
 * @returns Structured markdown text, or null if file is empty/unparseable
 */
export function parseConversationJsonl(absolutePath: string, projectDir: string): string | null {
  // Size guard
  const stat = statSync(absolutePath);
  if (stat.size > MAX_FILE_BYTES || stat.size === 0) {
    return null;
  }

  const raw = readFileSync(absolutePath, 'utf-8');
  const lines = raw.split('\n');

  let sessionId: string | undefined;
  let cwd: string | undefined;
  let timestamp: string | undefined;
  const segments: string[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    let rec: JsonlRecord;
    try {
      rec = JSON.parse(trimmed) as JsonlRecord;
    } catch {
      continue; // skip malformed lines
    }

    const recType = rec.type;
    if (!recType || SKIP_TYPES.has(recType)) continue;

    // Extract metadata from first record that has it
    if (!sessionId && rec.sessionId) sessionId = rec.sessionId;
    if (!cwd && rec.cwd) cwd = rec.cwd;
    if (!timestamp && rec.timestamp) timestamp = rec.timestamp;

    // Extract text from user messages
    if (recType === 'user') {
      const text = extractMessageText(rec.message);
      if (text) {
        segments.push(`[User]: ${text}`);
      }
      continue;
    }

    // Extract text from assistant messages
    if (recType === 'assistant') {
      const text = extractMessageText(rec.message);
      if (text) {
        segments.push(`[Assistant]: ${text}`);
      }
      continue;
    }

    // Extract text from summary records
    if (recType === 'summary') {
      const text = extractMessageText(rec.message);
      if (text) {
        segments.push(`[Summary]: ${text}`);
      }
      continue;
    }
  }

  if (segments.length === 0) {
    return null;
  }

  // Build structured header
  const header = [
    `# Session: ${sessionId ?? 'unknown'}`,
    `Project: ${projectDir}${cwd ? ` | CWD: ${cwd}` : ''}`,
    timestamp ? `Date: ${timestamp.slice(0, 10)}` : '',
  ].filter(Boolean).join('\n');

  return `${header}\n\n${segments.join('\n\n')}`;
}

/**
 * Extract plain text from a message's content field.
 * Handles both string content and array-of-blocks content.
 */
function extractMessageText(message: JsonlRecord['message']): string | null {
  if (!message?.content) return null;

  const content = message.content;

  // Simple string content (user messages)
  if (typeof content === 'string') {
    const trimmed = content.trim();
    return trimmed.length > 0 ? trimmed : null;
  }

  // Array of content blocks (assistant messages)
  if (Array.isArray(content)) {
    const textParts: string[] = [];
    for (const block of content) {
      if (typeof block !== 'object' || !block) continue;
      if (SKIP_BLOCK_TYPES.has(block.type)) continue;
      if (block.type === 'text' && typeof block.text === 'string') {
        const trimmed = block.text.trim();
        if (trimmed.length > 0) {
          textParts.push(trimmed);
        }
      }
    }
    return textParts.length > 0 ? textParts.join('\n\n') : null;
  }

  return null;
}

// --- Exchange-Level Parser ---

interface ToolUseBlock {
  type: 'tool_use';
  name?: string;
}

/**
 * Extract tool_use block names from an assistant message's content array.
 */
function extractToolCalls(content: string | ContentBlock[] | undefined): ToolCallInfo[] {
  if (!content || typeof content === 'string' || !Array.isArray(content)) return [];
  const tools: ToolCallInfo[] = [];
  for (const block of content) {
    if (typeof block === 'object' && block && block.type === 'tool_use') {
      const name = (block as unknown as ToolUseBlock).name;
      if (name) {
        tools.push({ toolName: name });
      }
    }
  }
  return tools;
}

/**
 * Parse a conversation JSONL file into structured exchanges (user+assistant pairs).
 * Each exchange groups a user message with its subsequent assistant response(s).
 * Tool calls are extracted from assistant messages.
 *
 * @param absolutePath - Full path to the .jsonl file
 * @returns Structured exchanges + metadata, or null if empty/unparseable
 */
export function parseConversationExchanges(absolutePath: string): {
  exchanges: ConversationExchange[];
  metadata: { sessionId?: string; cwd?: string; timestamp?: string };
} | null {
  const stat = statSync(absolutePath);
  if (stat.size > MAX_FILE_BYTES || stat.size === 0) {
    return null;
  }

  const raw = readFileSync(absolutePath, 'utf-8');
  const lines = raw.split('\n');

  let sessionId: string | undefined;
  let cwd: string | undefined;
  let timestamp: string | undefined;

  const exchanges: ConversationExchange[] = [];

  // Accumulator for current exchange being built
  let currentUserMsg = '';
  let currentUserLine = 0;
  let currentAssistantParts: string[] = [];
  let currentToolCalls: ToolCallInfo[] = [];
  let currentTimestamp: string | undefined;
  let lastLineNo = 0;
  let hasUser = false;

  function finalizeExchange(): void {
    if (!hasUser) return;
    const assistantMessage = currentAssistantParts.join('\n\n').trim();
    // Only emit exchanges that have at least a user message
    if (currentUserMsg.trim()) {
      exchanges.push({
        userMessage: currentUserMsg.trim(),
        assistantMessage,
        lineStart: currentUserLine,
        lineEnd: lastLineNo,
        timestamp: currentTimestamp,
        toolCalls: currentToolCalls,
      });
    }
    // Reset
    currentUserMsg = '';
    currentAssistantParts = [];
    currentToolCalls = [];
    currentTimestamp = undefined;
    hasUser = false;
  }

  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i]!.trim();
    if (!trimmed) continue;

    let rec: JsonlRecord;
    try {
      rec = JSON.parse(trimmed) as JsonlRecord;
    } catch {
      continue;
    }

    const recType = rec.type;
    if (!recType || SKIP_TYPES.has(recType)) continue;
    lastLineNo = i + 1;

    // Extract metadata from first record that has it
    if (!sessionId && rec.sessionId) sessionId = rec.sessionId;
    if (!cwd && rec.cwd) cwd = rec.cwd;
    if (!timestamp && rec.timestamp) timestamp = rec.timestamp;

    if (recType === 'user') {
      // Finalize previous exchange before starting new one
      finalizeExchange();
      currentUserMsg = extractMessageText(rec.message) ?? '';
      currentUserLine = i + 1;
      currentTimestamp = rec.timestamp;
      hasUser = true;
      continue;
    }

    if (recType === 'assistant') {
      const text = extractMessageText(rec.message);
      if (text) {
        currentAssistantParts.push(text);
      }
      // Extract tool calls from this assistant message
      if (rec.message?.content) {
        const tools = extractToolCalls(rec.message.content);
        currentToolCalls.push(...tools);
      }
      continue;
    }

    if (recType === 'summary') {
      const text = extractMessageText(rec.message);
      if (text) {
        currentAssistantParts.push(text);
      }
      continue;
    }
  }

  // Finalize last exchange
  finalizeExchange();

  if (exchanges.length === 0) {
    return null;
  }

  return {
    exchanges,
    metadata: { sessionId, cwd, timestamp },
  };
}

File: scripts/conversation_parser.py
"""
Python port of ~/claude-memory/src/conversation-parser.ts
Parses Claude Code conversation JSONL archives into structured text for graph ingestion.
"""

import json
import os
import sys
from dataclasses import dataclass, field
from typing import Optional

MAX_FILE_BYTES = 20 * 1024 * 1024  # 20MB
MAX_OUTPUT_BYTES = 50 * 1024  # 50KB cap per conversation

SKIP_TYPES = {'progress', 'queue-operation', 'file-history-snapshot'}
SKIP_BLOCK_TYPES = {'tool_use', 'tool_result', 'thinking'}


@dataclass
class ConversationExchange:
    user_message: str
    assistant_message: str
    timestamp: Optional[str] = None
    tool_names: list[str] = field(default_factory=list)


@dataclass
class ParseResult:
    exchanges: list[ConversationExchange]
    session_id: Optional[str] = None
    cwd: Optional[str] = None
    timestamp: Optional[str] = None


def extract_message_text(message: Optional[dict]) -> Optional[str]:
    """Extract plain text from a message's content field."""
    if not message or 'content' not in message:
        return None

    content = message['content']

    # Simple string content (user messages)
    if isinstance(content, str):
        trimmed = content.strip()
        return trimmed if trimmed else None

    # Array of content blocks (assistant messages)
    if isinstance(content, list):
        text_parts = []
        for block in content:
            if not isinstance(block, dict):
                continue
            if block.get('type') in SKIP_BLOCK_TYPES:
                continue
            if block.get('type') == 'text' and isinstance(block.get('text'), str):
                trimmed = block['text'].strip()
                if trimmed:
                    text_parts.append(trimmed)
        return '\n\n'.join(text_parts) if text_parts else None

    return None


def extract_tool_names(content) -> list[str]:
    """Extract tool_use block names from assistant message content."""
    if not content or isinstance(content, str) or not isinstance(content, list):
        return []
    tools = []
    for block in content:
        if isinstance(block, dict) and block.get('type') == 'tool_use':
            name = block.get('name')
            if name:
                tools.append(name)
    return tools


def parse_conversation_jsonl(filepath: str) -> Optional[ParseResult]:
    """
    Parse a conversation JSONL file into structured exchanges.
    Returns ParseResult or None if file is empty/unparseable.
    """
    try:
        size = os.path.getsize(filepath)
    except OSError:
        return None

    if size > MAX_FILE_BYTES or size == 0:
        return None

    session_id = None
    cwd = None
    timestamp = None

    exchanges: list[ConversationExchange] = []

    # Accumulator for current exchange
    current_user_msg = ''
    current_assistant_parts: list[str] = []
    current_tool_names: list[str] = []
    current_timestamp: Optional[str] = None
    has_user = False

    def finalize_exchange():
        nonlocal current_user_msg, current_assistant_parts, current_tool_names
        nonlocal current_timestamp, has_user

        if not has_user:
            return
        user_msg = current_user_msg.strip()
        if user_msg:
            assistant_msg = '\n\n'.join(current_assistant_parts).strip()
            exchanges.append(ConversationExchange(
                user_message=user_msg,
                assistant_message=assistant_msg,
                timestamp=current_timestamp,
                tool_names=current_tool_names,
            ))
        # Reset
        current_user_msg = ''
        current_assistant_parts = []
        current_tool_names = []
        current_timestamp = None
        has_user = False

    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            for line in f:
                trimmed = line.strip()
                if not trimmed:
                    continue

                try:
                    rec = json.loads(trimmed)
                except json.JSONDecodeError:
                    continue

                rec_type = rec.get('type')
                if not rec_type or rec_type in SKIP_TYPES:
                    continue

                # Extract metadata
                if not session_id and rec.get('sessionId'):
                    session_id = rec['sessionId']
                if not cwd and rec.get('cwd'):
                    cwd = rec['cwd']
                if not timestamp and rec.get('timestamp'):
                    timestamp = rec['timestamp']

                if rec_type == 'user':
                    finalize_exchange()
                    current_user_msg = extract_message_text(rec.get('message')) or ''
                    current_timestamp = rec.get('timestamp')
                    has_user = True
                    continue

                if rec_type == 'assistant':
                    text = extract_message_text(rec.get('message'))
                    if text:
                        current_assistant_parts.append(text)
                    # Extract tool names
                    msg = rec.get('message', {})
                    if msg.get('content'):
                        tools = extract_tool_names(msg['content'])
                        current_tool_names.extend(tools)
                    continue

                if rec_type == 'summary':
                    text = extract_message_text(rec.get('message'))
                    if text:
                        current_assistant_parts.append(text)
                    continue

        # Finalize last exchange
        finalize_exchange()

    except Exception as e:
        print(f"WARNING: Failed to parse {filepath}: {e}", file=sys.stderr)
        return None

    if not exchanges:
        return None

    return ParseResult(
        exchanges=exchanges,
        session_id=session_id,
        cwd=cwd,
        timestamp=timestamp,
    )


def format_for_graphiti(result: ParseResult, project_dir: str) -> str:
    """
    Format parsed conversation into text suitable for Graphiti episode ingestion.
    Caps output at MAX_OUTPUT_BYTES.
    """
    parts = []

    # Header
    header = f"Session: {result.session_id or 'unknown'}"
    header += f"\nProject: {project_dir}"
    if result.cwd:
        header += f" | CWD: {result.cwd}"
    if result.timestamp:
        header += f"\nDate: {result.timestamp[:10]}"
    parts.append(header)

    # Exchanges
    total_size = len(header)
    for ex in result.exchanges:
        exchange_text = f"\n[User]: {ex.user_message}"
        if ex.assistant_message:
            exchange_text += f"\n[Assistant]: {ex.assistant_message}"
        if ex.tool_names:
            unique_tools = list(dict.fromkeys(ex.tool_names))  # dedup, preserve order
            exchange_text += f"\n[Tools used]: {', '.join(unique_tools)}"

        total_size += len(exchange_text.encode('utf-8'))
        if total_size > MAX_OUTPUT_BYTES:
            parts.append("\n[... truncated due to size limit ...]")
            break
        parts.append(exchange_text)

    return '\n'.join(parts)
